File: central-brain/integration/openclaw-eth-client/examples/android/OpenClawEthClient.java
package com.example.openclaw;

import android.net.Network;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.Closeable;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.regex.Pattern;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;

/**
 * Business-neutral OpenClaw protocol 3 client bound to one Android Ethernet Network.
 *
 * <p>Callbacks run on OkHttp threads. The caller must switch to the main thread before updating
 * Android UI.
 */
public final class OpenClawEthClient implements Closeable {
    public interface Listener {
        void onStage(String stage);

        void onDelta(String text);
    }

    public static final class ImageInput {
        private static final Pattern FILE_NAME =
                Pattern.compile("[A-Za-z0-9][A-Za-z0-9._-]{0,95}");

        final String mimeType;
        final String fileName;
        final byte[] content;

        public ImageInput(String mimeType, String fileName, byte[] content) {
            if (!"image/png".equals(mimeType) && !"image/jpeg".equals(mimeType)) {
                throw new IllegalArgumentException("image MIME must be image/png or image/jpeg");
            }
            if (fileName == null || !FILE_NAME.matcher(fileName).matches()) {
                throw new IllegalArgumentException("image file name is invalid");
            }
            if (content == null || content.length == 0 || content.length > MAX_IMAGE_BYTES) {
                throw new IllegalArgumentException("image must contain 1..6291456 bytes");
            }
            requireImageSignature(mimeType, content);
            this.mimeType = mimeType;
            this.fileName = fileName;
            this.content = content.clone();
        }
    }

    private static final int PROTOCOL_VERSION = 3;
    private static final int MAX_IMAGE_BYTES = 6 * 1024 * 1024;
    private static final int MAX_TEXT_BYTES = 16 * 1024;
    private static final int MAX_PREAUTH_FRAME_BYTES = 65_536;
    private static final int MAX_MULTIMODAL_FRAME_BYTES = 8_500_000;
    private static final int MAX_INBOUND_FRAME_BYTES = 1_048_576;
    private static final Listener NOOP_LISTENER = new Listener() {
        @Override
        public void onStage(String stage) {
        }

        @Override
        public void onDelta(String text) {
        }
    };

    private final String endpoint;
    private final String origin;
    private final String token;
    private final int requestTimeoutMs;
    private final OkHttpClient httpClient;
    private final ScheduledExecutorService timeoutExecutor;
    private final AtomicBoolean closed = new AtomicBoolean();

    public OpenClawEthClient(
            Network ethernetNetwork,
            String endpoint,
            String token,
            int requestTimeoutMs) {
        Objects.requireNonNull(ethernetNetwork, "ethernetNetwork");
        this.endpoint = requireEndpoint(endpoint);
        this.origin = this.endpoint.substring(0, this.endpoint.length() - 1)
                .replaceFirst("^ws://", "http://");
        this.token = requireToken(token);
        if (requestTimeoutMs < 1_000 || requestTimeoutMs > 120_000) {
            throw new IllegalArgumentException("requestTimeoutMs must be in 1000..120000");
        }
        this.requestTimeoutMs = requestTimeoutMs;
        this.httpClient = new OkHttpClient.Builder()
                .socketFactory(ethernetNetwork.getSocketFactory())
                .connectTimeout(3, TimeUnit.SECONDS)
                .readTimeout(0, TimeUnit.MILLISECONDS)
                .pingInterval(15, TimeUnit.SECONDS)
                .build();
        ThreadFactory daemonFactory = runnable -> {
            Thread thread = new Thread(runnable, "openclaw-eth-timeout");
            thread.setDaemon(true);
            return thread;
        };
        this.timeoutExecutor = Executors.newSingleThreadScheduledExecutor(daemonFactory);
    }

    public CompletableFuture<String> queryText(String text, Listener listener) {
        return query(text, null, listener);
    }

    public CompletableFuture<String> queryTextAndImage(
            String text,
            ImageInput image,
            Listener listener) {
        return query(text, Objects.requireNonNull(image, "image"), listener);
    }

    private CompletableFuture<String> query(
            String text,
            ImageInput image,
            Listener listener) {
        if (closed.get()) {
            return CompletableFuture.failedFuture(
                    new IllegalStateException("OpenClaw client is closed"));
        }
        String boundedText = requireText(text);
        QuerySession session = new QuerySession(
                boundedText,
                image,
                listener == null ? NOOP_LISTENER : listener);
        Request request = new Request.Builder()
                .url(endpoint)
                .header("Origin", origin)
                .build();
        httpClient.newWebSocket(request, session);
        return session.future;
    }

    @Override
    public void close() {
        if (!closed.compareAndSet(false, true)) {
            return;
        }
        timeoutExecutor.shutdownNow();
        httpClient.dispatcher().cancelAll();
        httpClient.dispatcher().executorService().shutdown();
        httpClient.connectionPool().evictAll();
    }

    private final class QuerySession extends WebSocketListener {
        final String text;
        final ImageInput image;
        final Listener listener;
        final String sessionKey = "agent:main:eth-debug-"
                + UUID.randomUUID().toString().replace("-", "");
        final String idempotencyKey = UUID.randomUUID().toString();
        final CompletableFuture<String> future = new CompletableFuture<>();
        final AtomicBoolean terminal = new AtomicBoolean();
        final ScheduledFuture<?> timeout;

        WebSocket socket;
        String connectRequestId;
        String chatRequestId;
        String runId = "";
        String eventRunId = "";
        String streamedText = "";
        String finalText = "";
        boolean acknowledged;
        boolean finalSeen;

        QuerySession(String text, ImageInput image, Listener listener) {
            this.text = text;
            this.image = image;
            this.listener = listener;
            this.timeout = timeoutExecutor.schedule(
                    () -> fail(new TimeoutException("OpenClaw request deadline exceeded")),
                    requestTimeoutMs,
                    TimeUnit.MILLISECONDS);
        }

        @Override
        public synchronized void onOpen(WebSocket webSocket, Response response) {
            socket = webSocket;
            if (terminal.get()) {
                webSocket.cancel();
                return;
            }
            stage("WEBSOCKET_OPEN");
        }

        @Override
        public synchronized void onMessage(WebSocket webSocket, String raw) {
            if (terminal.get()) {
                return;
            }
            try {
                if (raw.getBytes(StandardCharsets.UTF_8).length > MAX_INBOUND_FRAME_BYTES) {
                    throw new IllegalStateException("inbound OpenClaw frame is too large");
                }
                JSONObject frame = new JSONObject(raw);
                if (isEvent(frame, "connect.challenge")) {
                    handleChallenge(frame);
                } else if (isResponse(frame, connectRequestId)) {
                    handleConnectResponse(frame);
                } else if (isResponse(frame, chatRequestId)) {
                    handleChatAck(frame);
                } else if (isEvent(frame, "chat")) {
                    handleChatEvent(frame);
                }
            } catch (RuntimeException | JSONException failure) {
                fail(failure);
            }
        }

        @Override
        public void onFailure(WebSocket webSocket, Throwable failure, Response response) {
            fail(new IllegalStateException("OpenClaw WebSocket failed", failure));
        }

        @Override
        public void onClosed(WebSocket webSocket, int code, String reason) {
            if (!terminal.get()) {
                fail(new IllegalStateException("OpenClaw WebSocket closed before final reply"));
            }
        }

        private void handleChallenge(JSONObject frame) throws JSONException {
            if (connectRequestId != null) {
                throw new IllegalStateException("duplicate OpenClaw challenge");
            }
            String nonce = frame.getJSONObject("payload").optString("nonce", "");
            if (nonce.isEmpty()) {
                throw new IllegalStateException("OpenClaw challenge nonce is missing");
            }
            stage("CHALLENGE_RECEIVED");
            connectRequestId = UUID.randomUUID().toString();
            JSONObject client = new JSONObject()
                    .put("id", "openclaw-control-ui")
                    .put("version", "eth-debug-client/1.0")
                    .put("platform", "android")
                    .put("mode", "webchat");
            JSONObject params = new JSONObject()
                    .put("minProtocol", PROTOCOL_VERSION)
                    .put("maxProtocol", PROTOCOL_VERSION)
                    .put("client", client)
                    .put("role", "operator")
                    .put("scopes", new JSONArray()
                            .put("operator.read")
                            .put("operator.write"))
                    .put("caps", new JSONArray())
                    .put("auth", new JSONObject().put("token", token))
                    .put("locale", "zh-CN")
                    .put("userAgent", "OpenClaw-Eth-Debug/1.0");
            send(rpcRequest(connectRequestId, "connect", params), MAX_PREAUTH_FRAME_BYTES);
            stage("CONNECT_SENT");
        }

        private void handleConnectResponse(JSONObject frame) throws JSONException {
            requireOk(frame, "OpenClaw authentication rejected");
            int protocol = frame.getJSONObject("payload").optInt("protocol", -1);
            if (protocol != PROTOCOL_VERSION) {
                throw new IllegalStateException("OpenClaw protocol mismatch: " + protocol);
            }
            stage("AUTHENTICATED");
            chatRequestId = UUID.randomUUID().toString();
            JSONObject params = new JSONObject()
                    .put("sessionKey", sessionKey)
                    .put("message", text)
                    .put("deliver", false)
                    .put("idempotencyKey", idempotencyKey);
            int maximumBytes = MAX_PREAUTH_FRAME_BYTES;
            if (image != null) {
                JSONObject attachment = new JSONObject()
                        .put("type", "image")
                        .put("mimeType", image.mimeType)
                        .put("fileName", image.fileName)
                        .put("content", Base64.getEncoder().encodeToString(image.content));
                params.put("attachments", new JSONArray().put(attachment));
                maximumBytes = MAX_MULTIMODAL_FRAME_BYTES;
            }
            send(rpcRequest(chatRequestId, "chat.send", params), maximumBytes);
            stage("CHAT_SENT");
        }

        private void handleChatAck(JSONObject frame) throws JSONException {
            requireOk(frame, "OpenClaw chat request rejected");
            acknowledged = true;
            JSONObject payload = frame.optJSONObject("payload");
            runId = payload == null ? "" : payload.optString("runId", "");
            if (!runId.isEmpty() && !eventRunId.isEmpty() && !runId.equals(eventRunId)) {
                throw new IllegalStateException("OpenClaw runId changed during request");
            }
            stage("CHAT_ACKNOWLEDGED");
            completeIfReady();
        }

        private void handleChatEvent(JSONObject frame) throws JSONException {
            JSONObject payload = frame.optJSONObject("payload");
            if (payload == null || !sessionKey.equals(payload.optString("sessionKey", ""))) {
                return;
            }
            String candidateRunId = payload.optString("runId", "");
            if (!candidateRunId.isEmpty()) {
                if (!runId.isEmpty() && !runId.equals(candidateRunId)) {
                    return;
                }
                if (!eventRunId.isEmpty() && !eventRunId.equals(candidateRunId)) {
                    return;
                }
                eventRunId = candidateRunId;
            }
            String state = payload.optString("state", "");
            if ("delta".equals(state)) {
                streamedText = mergeText(streamedText, extractText(payload.opt("message")));
                if (!streamedText.isEmpty()) {
                    listener.onDelta(streamedText);
                }
            } else if ("final".equals(state)) {
                finalSeen = true;
                String terminalText = extractText(payload.opt("message"));
                finalText = terminalText.isEmpty() ? streamedText : terminalText;
                stage("FINAL_RECEIVED");
                completeIfReady();
            } else if ("error".equals(state)) {
                throw new IllegalStateException("OpenClaw returned a terminal chat error");
            }
        }

        private void completeIfReady() {
            if (!acknowledged || !finalSeen) {
                return;
            }
            if (finalText.isEmpty()) {
                fail(new IllegalStateException("OpenClaw final reply is empty"));
                return;
            }
            if (!terminal.compareAndSet(false, true)) {
                return;
            }
            timeout.cancel(false);
            future.complete(finalText);
            if (socket != null) {
                socket.close(1000, "complete");
            }
        }

        private void send(JSONObject frame, int maximumBytes) {
            if (socket == null) {
                throw new IllegalStateException("OpenClaw WebSocket is unavailable");
            }
            String encoded = frame.toString();
            int bytes = encoded.getBytes(StandardCharsets.UTF_8).length;
            if (bytes == 0 || bytes > maximumBytes) {
                throw new IllegalStateException("OpenClaw outbound frame is too large");
            }
            if (!socket.send(encoded)) {
                throw new IllegalStateException("OpenClaw outbound queue is closed");
            }
        }

        private void stage(String stage) {
            listener.onStage(stage);
        }

        private void fail(Throwable failure) {
            if (!terminal.compareAndSet(false, true)) {
                return;
            }
            timeout.cancel(false);
            future.completeExceptionally(failure);
            WebSocket current = socket;
            if (current != null) {
                current.cancel();
            }
        }
    }

    private static JSONObject rpcRequest(String id, String method, JSONObject params)
            throws JSONException {
        return new JSONObject()
                .put("type", "req")
                .put("id", id)
                .put("method", method)
                .put("params", params);
    }

    private static boolean isResponse(JSONObject frame, String id) {
        return id != null
                && "res".equals(frame.optString("type", ""))
                && id.equals(frame.optString("id", ""));
    }

    private static boolean isEvent(JSONObject frame, String event) {
        return "event".equals(frame.optString("type", ""))
                && event.equals(frame.optString("event", ""));
    }

    private static void requireOk(JSONObject frame, String prefix) {
        if (frame.optBoolean("ok", false)) {
            return;
        }
        JSONObject error = frame.optJSONObject("error");
        String code = error == null ? "UNKNOWN" : safeError(error.optString("code", "UNKNOWN"));
        String message = error == null ? "" : safeError(error.optString("message", ""));
        throw new IllegalStateException(prefix + " code=" + code + " message=" + message);
    }

    private static String extractText(Object message) throws JSONException {
        if (message == null || message == JSONObject.NULL) {
            return "";
        }
        if (message instanceof String) {
            return (String) message;
        }
        if (!(message instanceof JSONObject)) {
            return "";
        }
        JSONObject object = (JSONObject) message;
        if (object.opt("text") instanceof String) {
            return object.getString("text");
        }
        Object content = object.opt("content");
        if (content instanceof String) {
            return (String) content;
        }
        if (!(content instanceof JSONArray)) {
            return "";
        }
        JSONArray parts = (JSONArray) content;
        StringBuilder value = new StringBuilder();
        for (int index = 0; index < parts.length(); index++) {
            JSONObject part = parts.optJSONObject(index);
            if (part != null
                    && "text".equals(part.optString("type", ""))
                    && part.opt("text") instanceof String) {
                value.append(part.getString("text"));
            }
        }
        return value.toString();
    }

    private static String mergeText(String accumulated, String incoming) {
        if (incoming == null || incoming.isEmpty()) {
            return accumulated;
        }
        if (incoming.startsWith(accumulated)) {
            return incoming;
        }
        if (accumulated.endsWith(incoming)) {
            return accumulated;
        }
        return accumulated + incoming;
    }

    private static String requireEndpoint(String value) {
        if (!"ws://169.254.208.110:18789/".equals(value)) {
            throw new IllegalArgumentException(
                    "endpoint must be ws://169.254.208.110:18789/");
        }
        return value;
    }

    private static String requireToken(String value) {
        if (value == null || value.length() < 8 || value.length() > 256) {
            throw new IllegalArgumentException("OpenClaw token is invalid");
        }
        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            if (character < 0x21 || character > 0x7e) {
                throw new IllegalArgumentException("OpenClaw token is invalid");
            }
        }
        return value;
    }

    private static String requireText(String value) {
        if (value == null || value.trim().isEmpty()) {
            throw new IllegalArgumentException("text query is empty");
        }
        if (value.getBytes(StandardCharsets.UTF_8).length > MAX_TEXT_BYTES) {
            throw new IllegalArgumentException("text query exceeds 16 KiB");
        }
        return value;
    }

    private static String safeError(String value) {
        if (value == null) {
            return "";
        }
        String sanitized = value.replaceAll("[^A-Za-z0-9_.: -]", "_");
        return sanitized.length() <= 160 ? sanitized : sanitized.substring(0, 160);
    }

    private static void requireImageSignature(String mimeType, byte[] content) {
        boolean png = content.length >= 8
                && (content[0] & 0xff) == 0x89
                && content[1] == 0x50
                && content[2] == 0x4e
                && content[3] == 0x47
                && content[4] == 0x0d
                && content[5] == 0x0a
                && content[6] == 0x1a
                && content[7] == 0x0a;
        boolean jpeg = content.length >= 4
                && (content[0] & 0xff) == 0xff
                && (content[1] & 0xff) == 0xd8
                && (content[content.length - 2] & 0xff) == 0xff
                && (content[content.length - 1] & 0xff) == 0xd9;
        if (("image/png".equals(mimeType) && !png)
                || ("image/jpeg".equals(mimeType) && !jpeg)) {
            throw new IllegalArgumentException("image signature does not match MIME");
        }
    }
}

File: central-brain/integration/openclaw-eth-client/examples/linux/openclaw_eth_client.py
#!/usr/bin/env python3
"""Business-neutral OpenClaw protocol 3 client for a directly attached ETH node."""

from __future__ import annotations

import argparse
import base64
import json
import mimetypes
import os
import re
import sys
import time
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any
from urllib.parse import urlparse

import websocket


PROTOCOL_VERSION = 3
DEFAULT_ENDPOINT = "ws://169.254.208.110:18789/"
MAX_TEXT_BYTES = 16 * 1024
MAX_IMAGE_BYTES = 6 * 1024 * 1024
MAX_PREAUTH_FRAME_BYTES = 65_536
MAX_MULTIMODAL_FRAME_BYTES = 8_500_000
MAX_INBOUND_FRAME_BYTES = 1_048_576
FILE_NAME_PATTERN = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,95}$")


class OpenClawError(RuntimeError):
    """Bounded protocol or transport failure."""


@dataclass(frozen=True)
class ImageInput:
    mime_type: str
    file_name: str
    content: bytes

    @classmethod
    def from_path(cls, path: Path) -> "ImageInput":
        if not path.is_file():
            raise OpenClawError(f"image does not exist: {path}")
        size = path.stat().st_size
        if size < 1 or size > MAX_IMAGE_BYTES:
            raise OpenClawError("image must contain 1..6291456 bytes")
        content = path.read_bytes()
        mime_type = mimetypes.guess_type(path.name)[0]
        if mime_type not in ("image/png", "image/jpeg"):
            raise OpenClawError("image extension must identify PNG or JPEG")
        file_name = path.name
        if not FILE_NAME_PATTERN.fullmatch(file_name):
            raise OpenClawError("image file name is invalid")
        _require_image_signature(mime_type, content)
        return cls(mime_type=mime_type, file_name=file_name, content=content)


class OpenClawEthClient:
    def __init__(
        self,
        endpoint: str,
        token: str,
        connect_timeout_s: float = 3.0,
        request_timeout_s: float = 120.0,
    ) -> None:
        parsed = urlparse(endpoint)
        if (
            parsed.scheme != "ws"
            or parsed.hostname != "169.254.208.110"
            or parsed.port != 18789
            or parsed.path != "/"
            or parsed.params
            or parsed.query
            or parsed.fragment
        ):
            raise OpenClawError(
                "endpoint must be ws://169.254.208.110:18789/"
            )
        if (
            not isinstance(token, str)
            or not 8 <= len(token) <= 256
            or any(ord(character) < 0x21 or ord(character) > 0x7E for character in token)
        ):
            raise OpenClawError("OpenClaw token is invalid")
        if connect_timeout_s <= 0:
            raise OpenClawError("connect timeout must be positive")
        if request_timeout_s <= 0 or request_timeout_s > 120:
            raise OpenClawError("request timeout must be in (0, 120]")
        self.endpoint = endpoint
        self.origin = f"http://{parsed.hostname}:{parsed.port}"
        self.token = token
        self.connect_timeout_s = connect_timeout_s
        self.request_timeout_s = request_timeout_s

    def query(self, text: str, image: ImageInput | None = None) -> str:
        text = _require_text(text)
        deadline = time.monotonic() + self.request_timeout_s
        host = urlparse(self.endpoint).hostname
        ws = websocket.create_connection(
            self.endpoint,
            timeout=self.connect_timeout_s,
            origin=self.origin,
            http_no_proxy=[host],
            enable_multithread=True,
        )
        try:
            _stage("WEBSOCKET_OPEN")
            ws.settimeout(_remaining(deadline))
            challenge = _receive_json(ws, deadline)
            if not _is_event(challenge, "connect.challenge"):
                raise OpenClawError("OpenClaw connect challenge is missing")
            nonce = _object(challenge, "payload").get("nonce")
            if not isinstance(nonce, str) or not nonce:
                raise OpenClawError("OpenClaw challenge nonce is missing")
            _stage("CHALLENGE_RECEIVED")

            connect_id = str(uuid.uuid4())
            connect_params = {
                "minProtocol": PROTOCOL_VERSION,
                "maxProtocol": PROTOCOL_VERSION,
                "client": {
                    "id": "openclaw-control-ui",
                    "version": "eth-debug-client/1.0",
                    "platform": "linux",
                    "mode": "webchat",
                },
                "role": "operator",
                "scopes": ["operator.read", "operator.write"],
                "caps": [],
                "auth": {"token": self.token},
                "locale": "zh-CN",
                "userAgent": "OpenClaw-Eth-Debug/1.0",
            }
            _send_json(
                ws,
                _rpc_request(connect_id, "connect", connect_params),
                MAX_PREAUTH_FRAME_BYTES,
            )
            _stage("CONNECT_SENT")
            connect_response = _read_matching_response(ws, connect_id, deadline)
            _require_ok(connect_response, "OpenClaw authentication rejected")
            protocol = _object(connect_response, "payload").get("protocol")
            if protocol != PROTOCOL_VERSION:
                raise OpenClawError(f"OpenClaw protocol mismatch: {protocol!r}")
            _stage("AUTHENTICATED")

            session_key = f"agent:main:eth-debug-{uuid.uuid4().hex}"
            idempotency_key = str(uuid.uuid4())
            chat_id = str(uuid.uuid4())
            chat_params: dict[str, Any] = {
                "sessionKey": session_key,
                "message": text,
                "deliver": False,
                "idempotencyKey": idempotency_key,
            }
            frame_limit = MAX_PREAUTH_FRAME_BYTES
            if image is not None:
                chat_params["attachments"] = [
                    {
                        "type": "image",
                        "mimeType": image.mime_type,
                        "fileName": image.file_name,
                        "content": base64.b64encode(image.content).decode("ascii"),
                    }
                ]
                frame_limit = MAX_MULTIMODAL_FRAME_BYTES
            _send_json(
                ws,
                _rpc_request(chat_id, "chat.send", chat_params),
                frame_limit,
            )
            _stage("CHAT_SENT")
            return self._read_chat(ws, chat_id, session_key, deadline)
        except websocket.WebSocketTimeoutException as failure:
            raise OpenClawError("OpenClaw request deadline exceeded") from failure
        except websocket.WebSocketException as failure:
            raise OpenClawError("OpenClaw WebSocket transport failed") from failure
        finally:
            try:
                ws.close(status=1000, reason="complete")
            except websocket.WebSocketException:
                pass

    def _read_chat(
        self,
        ws: websocket.WebSocket,
        chat_id: str,
        session_key: str,
        deadline: float,
    ) -> str:
        acknowledged = False
        final_seen = False
        run_id = ""
        event_run_id = ""
        streamed_text = ""
        final_text = ""

        while True:
            frame = _receive_json(ws, deadline)
            if _is_response(frame, chat_id):
                _require_ok(frame, "OpenClaw chat request rejected")
                acknowledged = True
                payload = frame.get("payload")
                if isinstance(payload, dict) and isinstance(payload.get("runId"), str):
                    run_id = payload["runId"]
                if run_id and event_run_id and run_id != event_run_id:
                    raise OpenClawError("OpenClaw runId changed during request")
                _stage("CHAT_ACKNOWLEDGED")
            elif _is_event(frame, "chat"):
                payload = frame.get("payload")
                if not isinstance(payload, dict):
                    continue
                if payload.get("sessionKey") != session_key:
                    continue
                candidate_run_id = payload.get("runId")
                if isinstance(candidate_run_id, str) and candidate_run_id:
                    if run_id and candidate_run_id != run_id:
                        continue
                    if event_run_id and candidate_run_id != event_run_id:
                        continue
                    event_run_id = candidate_run_id
                state = payload.get("state")
                if state == "delta":
                    streamed_text = _merge_text(
                        streamed_text, _extract_text(payload.get("message"))
                    )
                    if streamed_text:
                        _stage(f"DELTA chars={len(streamed_text)}")
                elif state == "final":
                    final_seen = True
                    terminal_text = _extract_text(payload.get("message"))
                    final_text = terminal_text or streamed_text
                    _stage("FINAL_RECEIVED")
                elif state == "error":
                    raise OpenClawError("OpenClaw returned a terminal chat error")

            if acknowledged and final_seen:
                if not final_text:
                    raise OpenClawError("OpenClaw final reply is empty")
                return final_text


def _read_matching_response(
    ws: websocket.WebSocket, request_id: str, deadline: float
) -> dict[str, Any]:
    while True:
        frame = _receive_json(ws, deadline)
        if _is_response(frame, request_id):
            return frame


def _send_json(
    ws: websocket.WebSocket, value: dict[str, Any], maximum_bytes: int
) -> None:
    encoded = json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode(
        "utf-8"
    )
    if not encoded or len(encoded) > maximum_bytes:
        raise OpenClawError("OpenClaw outbound frame is too large")
    ws.send(encoded.decode("utf-8"))


def _receive_json(
    ws: websocket.WebSocket, deadline: float
) -> dict[str, Any]:
    ws.settimeout(_remaining(deadline))
    raw = ws.recv()
    if isinstance(raw, bytes):
        try:
            raw = raw.decode("utf-8", errors="strict")
        except UnicodeDecodeError as failure:
            raise OpenClawError("OpenClaw frame is not valid UTF-8") from failure
    if not isinstance(raw, str):
        raise OpenClawError("OpenClaw returned a non-text frame")
    if len(raw.encode("utf-8")) > MAX_INBOUND_FRAME_BYTES:
        raise OpenClawError("OpenClaw inbound frame is too large")
    try:
        value = json.loads(raw)
    except json.JSONDecodeError as failure:
        raise OpenClawError("OpenClaw returned invalid JSON") from failure
    if not isinstance(value, dict):
        raise OpenClawError("OpenClaw frame must be a JSON object")
    return value


def _rpc_request(
    request_id: str, method: str, params: dict[str, Any]
) -> dict[str, Any]:
    return {
        "type": "req",
        "id": request_id,
        "method": method,
        "params": params,
    }


def _is_response(frame: dict[str, Any], request_id: str) -> bool:
    return frame.get("type") == "res" and frame.get("id") == request_id


def _is_event(frame: dict[str, Any], event: str) -> bool:
    return frame.get("type") == "event" and frame.get("event") == event


def _require_ok(frame: dict[str, Any], prefix: str) -> None:
    if frame.get("ok") is True:
        return
    error = frame.get("error")
    code = _safe_error(error.get("code")) if isinstance(error, dict) else "UNKNOWN"
    message = _safe_error(error.get("message")) if isinstance(error, dict) else ""
    raise OpenClawError(f"{prefix} code={code} message={message}")


def _object(value: dict[str, Any], field: str) -> dict[str, Any]:
    child = value.get(field)
    if not isinstance(child, dict):
        raise OpenClawError(f"OpenClaw field is not an object: {field}")
    return child


def _extract_text(message: Any) -> str:
    if isinstance(message, str):
        return message
    if not isinstance(message, dict):
        return ""
    if isinstance(message.get("text"), str):
        return message["text"]
    content = message.get("content")
    if isinstance(content, str):
        return content
    if not isinstance(content, list):
        return ""
    return "".join(
        part["text"]
        for part in content
        if isinstance(part, dict)
        and part.get("type") == "text"
        and isinstance(part.get("text"), str)
    )


def _merge_text(accumulated: str, incoming: str) -> str:
    if not incoming:
        return accumulated
    if incoming.startswith(accumulated):
        return incoming
    if accumulated.endswith(incoming):
        return accumulated
    return accumulated + incoming


def _require_text(value: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise OpenClawError("text query is empty")
    if len(value.encode("utf-8")) > MAX_TEXT_BYTES:
        raise OpenClawError("text query exceeds 16 KiB")
    return value


def _require_image_signature(mime_type: str, content: bytes) -> None:
    png = content.startswith(b"\x89PNG\r\n\x1a\n")
    jpeg = (
        len(content) >= 4
        and content.startswith(b"\xff\xd8")
        and content.endswith(b"\xff\xd9")
    )
    if (mime_type == "image/png" and not png) or (
        mime_type == "image/jpeg" and not jpeg
    ):
        raise OpenClawError("image signature does not match MIME")


def _remaining(deadline: float) -> float:
    remaining = deadline - time.monotonic()
    if remaining <= 0:
        raise OpenClawError("OpenClaw request deadline exceeded")
    return remaining


def _safe_error(value: Any) -> str:
    if not isinstance(value, str):
        return "UNKNOWN"
    return re.sub(r"[^A-Za-z0-9_.: -]", "_", value)[:160]


def _stage(value: str) -> None:
    print(f"openclaw_stage={value}", file=sys.stderr, flush=True)


def _parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Send a text or text+image query to OpenClaw over Ethernet."
    )
    parser.add_argument("--endpoint", default=DEFAULT_ENDPOINT)
    parser.add_argument("--text", required=True)
    parser.add_argument("--image", type=Path)
    parser.add_argument("--timeout", type=float, default=120.0)
    return parser.parse_args()


def main() -> int:
    args = _parse_args()
    token = os.environ.get("OPENCLAW_TOKEN")
    if not token:
        print("OPENCLAW_TOKEN is not set", file=sys.stderr)
        return 2
    try:
        image = ImageInput.from_path(args.image) if args.image else None
        client = OpenClawEthClient(
            endpoint=args.endpoint,
            token=token,
            request_timeout_s=args.timeout,
        )
        reply = client.query(args.text, image)
    except OpenClawError as failure:
        print(f"openclaw_error={failure}", file=sys.stderr)
        return 1
    print(reply)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
